### PR TITLE
Qualification tool: Parse expressions in FilterExec

### PR DESCRIPTION
**Pull request by nartal1**
This contributes to https://github.com/NVIDIA/spark-rapids/issues/5617 . 
In this PR we parse expressions which are in FilterExec. It supports NOT/AND/OR within the Exec. With this PR, if there are expressions which are not supported by the plugin, then FilterExec is updated as NOT SUPPORTED in the plugin. This PR doesn't update the scores based on expressions. 
Added test for non-supported expression. For supported ones, the current tests should be sufficient.

### CODE REVIEW

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala [hunk 1]
```diff
@@ -249,4 +249,88 @@ object SQLPlanParser extends Logging {
...
+            logWarning(s"predicate string is $predStr")
+            result += predStr
+          // TODO - lookup function name
+          case None => logWarning(s"Incorrect expression - $expr")
+        }
+      } else {
+        // likely some function call
+        functionPattern.findFirstMatchIn(expr) match {
+          case Some(func) =>
+            logWarning(s" found func: $func")
```
Comment (tgravescs) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala:323]: change the logWarnings to Debug for now

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala [hunk 2]
```diff
@@ -249,4 +249,88 @@ object SQLPlanParser extends Logging {
...
+        // likely some conditional expression
+        // TODO - add in arithmetic stuff (- / * )
+        conditionalExprPattern.findFirstMatchIn(expr) match {
+          case Some(func) =>
+            logWarning(s" found expr: $func")
+            if (func.groupCount < 3) {
+              logError(s"found incomplete expression - $func")
+            }
+            val lhs = func.group(1)
+            functionPattern.findFirstMatchIn(lhs) match {
```
Comment (tgravescs) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala:292]: I think this should be the same logic as the else below. make common function

**Review by tgravescs — COMMENTED** [review]
overall looks like a good start, I think we can adjust as we add more.

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala (hunk 1 above)
Comment (nartal1) [reply tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala]: done

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala (hunk 2 above)
Comment (nartal1) [reply tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala]: created common function

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala [hunk 3]
```diff
@@ -249,4 +249,88 @@ object SQLPlanParser extends Logging {
...
+              case Some(func) =>
+                val func1 = func.group(1)
+                if (!func1.toLowerCase.equals("cast")) { // cast should not be added to the result
```
[the remark this answers is not in the document]
Comment (nartal1) [reply tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala]: added comments. Please let me know if I need to elaborate more.

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/FilterExecParser.scala [hunk 4]
```diff
@@ -30,7 +30,11 @@ case class FilterExecParser(
   override def parse: ExecInfo = {
     // filter doesn't have duration
     val duration = None
-    val (speedupFactor, isSupported) = if (checker.isExecSupported(fullExecName)) {
+    val exprString = node.desc.replaceFirst("Filter ", "")
+    val expressions = SQLPlanParser.parseFilterExpressions(exprString).map(_.toLowerCase)
```
[the remark this answers is not in the document]
Comment (nartal1) [reply tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/FilterExecParser.scala]: done

**Comment by nartal1** [comment]
build

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala [hunk 5]
```diff
@@ -249,4 +250,95 @@ object SQLPlanParser extends Logging {
...
   }
 
+  private def getFunctionName(functionPattern: Regex, expr:String): Option[String] = {
+    val funcName = functionPattern.findFirstMatchIn(expr) match {
+      case Some(func) =>
+        val func1 = func.group(1)
+        // `cast` is not an expression hence should be ignored. In the physical plan cast is
+        // usually presented as function call for example: `cast(value#9 as date)`. We add
+        // other function names to the result.
+          if (!func1.equalsIgnoreCase("cast")) {
```
Comment (tgravescs) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala:260]: spacing off here

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeChecker.scala [hunk 6]
```diff
@@ -250,4 +250,20 @@ class PluginTypeChecker extends Logging {
       false
     }
   }
+
+  def isExprSupported(expr: String): Boolean = {
+    val exprLowecase = expr.toLowerCase
```
Comment (tgravescs) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/qualification/PluginTypeChecker.scala:255]: nit exprLowerCase

**Comment by nartal1** [comment]
build

**Pull request merged**